File: backend/internal/domain/workvalues/bradley_terry.go
package workvalues

import "math"

const (
	sigmaSq = 3.0
	maxIter = 100
	tol     = 1e-6
)

type BTResult struct {
	Mu []float64
	SE []float64
}

func EstimateBT(wins [N][N]int) BTResult {
	mu := make([]float64, N)

	for range maxIter {
		grad := make([]float64, N)
		H := make([][]float64, N)
		for i := range H {
			H[i] = make([]float64, N)
		}

		for i := range N {
			for j := i + 1; j < N; j++ {
				nij := wins[i][j] + wins[j][i]
				if nij == 0 {
					continue
				}
				pij := sigmoid(mu[i] - mu[j])
				w := pij * (1 - pij)

				grad[i] += float64(wins[i][j]) - float64(nij)*pij
				grad[j] -= float64(wins[i][j]) - float64(nij)*pij

				H[i][i] -= float64(nij) * w
				H[j][j] -= float64(nij) * w
				H[i][j] += float64(nij) * w
				H[j][i] += float64(nij) * w
			}
			grad[i] -= mu[i] / sigmaSq
			H[i][i] -= 1.0 / sigmaSq
		}

		delta := solveLinear(H, grad)
		if delta == nil {
			break
		}

		maxDelta := 0.0
		for i := range N {
			mu[i] -= delta[i]
			if d := math.Abs(delta[i]); d > maxDelta {
				maxDelta = d
			}
		}

		mean := 0.0
		for _, v := range mu {
			mean += v
		}
		mean /= float64(N)
		for i := range mu {
			mu[i] -= mean
		}

		if maxDelta < tol {
			break
		}
	}

	se := computeSE(mu, wins)
	return BTResult{Mu: mu, SE: se}
}

func computeSE(mu []float64, wins [N][N]int) []float64 {
	H := make([][]float64, N)
	for i := range H {
		H[i] = make([]float64, N)
	}

	for i := range N {
		for j := i + 1; j < N; j++ {
			nij := wins[i][j] + wins[j][i]
			if nij == 0 {
				continue
			}
			pij := sigmoid(mu[i] - mu[j])
			w := float64(nij) * pij * (1 - pij)
			H[i][i] += w
			H[j][j] += w
			H[i][j] -= w
			H[j][i] -= w
		}
		H[i][i] += 1.0 / sigmaSq
	}

	inv := invertMatrix(H)
	if inv == nil {
		se := make([]float64, N)
		for i := range se {
			se[i] = math.Inf(1)
		}
		return se
	}

	se := make([]float64, N)
	for i := range N {
		v := inv[i][i]
		if v < 0 {
			v = 0
		}
		se[i] = math.Sqrt(v)
	}
	return se
}

func sigmoid(x float64) float64 {
	return 1.0 / (1.0 + math.Exp(-x))
}

func solveLinear(hess [][]float64, b []float64) []float64 {
	n := len(hess)
	A := make([][]float64, n)
	for i := range A {
		row := make([]float64, n+1)
		for j := range n {
			row[j] = -hess[i][j]
		}
		row[n] = -b[i]
		A[i] = row
	}

	for col := range n {
		pivotRow := col
		pivotVal := math.Abs(A[col][col])
		for row := col + 1; row < n; row++ {
			if v := math.Abs(A[row][col]); v > pivotVal {
				pivotRow = row
				pivotVal = v
			}
		}
		if pivotVal < 1e-12 {
			return nil
		}
		if pivotRow != col {
			A[col], A[pivotRow] = A[pivotRow], A[col]
		}
		pivot := A[col][col]
		for j := col; j <= n; j++ {
			A[col][j] /= pivot
		}
		for row := range n {
			if row == col {
				continue
			}
			factor := A[row][col]
			for j := col; j <= n; j++ {
				A[row][j] -= factor * A[col][j]
			}
		}
	}

	result := make([]float64, n)
	for i := range result {
		result[i] = A[i][n]
	}
	return result
}

func invertMatrix(mat [][]float64) [][]float64 {
	n := len(mat)
	aug := make([][]float64, n)
	for i := range aug {
		row := make([]float64, 2*n)
		copy(row, mat[i])
		row[n+i] = 1
		aug[i] = row
	}

	for col := range n {
		pivotRow := col
		pivotVal := math.Abs(aug[col][col])
		for row := col + 1; row < n; row++ {
			if v := math.Abs(aug[row][col]); v > pivotVal {
				pivotRow = row
				pivotVal = v
			}
		}
		if pivotVal < 1e-12 {
			return nil
		}
		if pivotRow != col {
			aug[col], aug[pivotRow] = aug[pivotRow], aug[col]
		}
		pivot := aug[col][col]
		for j := range 2 * n {
			aug[col][j] /= pivot
		}
		for row := range n {
			if row == col {
				continue
			}
			factor := aug[row][col]
			for j := range 2 * n {
				aug[row][j] -= factor * aug[col][j]
			}
		}
	}

	result := make([][]float64, n)
	for i := range result {
		result[i] = aug[i][n:]
	}
	return result
}

File: frontend/src/features/work-values/lib/bradley-terry.ts
import { N } from "./needs";

const SIGMA_SQ = 3.0;
const MAX_ITER = 100;
const TOL = 1e-6;

function sigmoid(x: number): number {
  return 1 / (1 + Math.exp(-x));
}

export interface BTResult {
  mu: number[];
  se: number[];
}

/**
 * Bradley-Terry モデルを Newton-Raphson 法で推定する。
 * wins[i][j] = ニーズ i が j に勝った回数。
 */
export function estimateBT(wins: number[][]): BTResult {
  const mu = new Array(N).fill(0);

  for (let iter = 0; iter < MAX_ITER; iter++) {
    const grad = new Array(N).fill(0);
    const H = Array.from({ length: N }, () => new Array(N).fill(0));

    for (let i = 0; i < N; i++) {
      for (let j = i + 1; j < N; j++) {
        const nij = wins[i][j] + wins[j][i];
        if (nij === 0) continue;

        const pij = sigmoid(mu[i] - mu[j]);
        const w = pij * (1 - pij);

        grad[i] += wins[i][j] - nij * pij;
        grad[j] -= wins[i][j] - nij * pij;

        H[i][i] -= nij * w;
        H[j][j] -= nij * w;
        H[i][j] += nij * w;
        H[j][i] += nij * w;
      }

      grad[i] -= mu[i] / SIGMA_SQ;
      H[i][i] -= 1 / SIGMA_SQ;
    }

    const delta = solveLinear(H, grad);
    if (!delta) break;

    let maxDelta = 0;
    for (let i = 0; i < N; i++) {
      mu[i] -= delta[i];
      maxDelta = Math.max(maxDelta, Math.abs(delta[i]));
    }

    const mean = mu.reduce((s, v) => s + v, 0) / N;
    for (let i = 0; i < N; i++) mu[i] -= mean;

    if (maxDelta < TOL) break;
  }

  const se = computeSE(mu, wins);

  return { mu, se };
}

/**
 * ヘッセ行列の逆行列の対角要素から SE を計算する。
 */
function computeSE(mu: number[], wins: number[][]): number[] {
  const H = Array.from({ length: N }, () => new Array(N).fill(0));

  for (let i = 0; i < N; i++) {
    for (let j = i + 1; j < N; j++) {
      const nij = wins[i][j] + wins[j][i];
      if (nij === 0) continue;

      const pij = sigmoid(mu[i] - mu[j]);
      const w = nij * pij * (1 - pij);

      H[i][i] += w;
      H[j][j] += w;
      H[i][j] -= w;
      H[j][i] -= w;
    }
    H[i][i] += 1 / SIGMA_SQ;
  }

  const invH = invertMatrix(H);
  if (!invH) return new Array(N).fill(Infinity);

  return invH.map((row, i) => Math.sqrt(Math.max(0, row[i])));
}

/**
 * H * x = b を解く（ガウス消去法）。
 * H は負定値なので -H * x = -b として解く。
 */
function solveLinear(H: number[][], b: number[]): number[] | null {
  const n = H.length;
  const A = H.map((row, i) => [...row.map((v) => -v), -b[i]]);

  for (let col = 0; col < n; col++) {
    let pivotRow = col;
    let pivotVal = Math.abs(A[col][col]);
    for (let row = col + 1; row < n; row++) {
      if (Math.abs(A[row][col]) > pivotVal) {
        pivotRow = row;
        pivotVal = Math.abs(A[row][col]);
      }
    }
    if (pivotVal < 1e-12) return null;

    if (pivotRow !== col) {
      [A[col], A[pivotRow]] = [A[pivotRow], A[col]];
    }

    const pivot = A[col][col];
    for (let j = col; j <= n; j++) A[col][j] /= pivot;

    for (let row = 0; row < n; row++) {
      if (row === col) continue;
      const factor = A[row][col];
      for (let j = col; j <= n; j++) {
        A[row][j] -= factor * A[col][j];
      }
    }
  }

  return A.map((row) => row[n]);
}

/**
 * 行列の逆行列を計算する（ガウス・ジョルダン法）。
 */
function invertMatrix(M: number[][]): number[][] | null {
  const n = M.length;
  const aug = M.map((row, i) => {
    const ext = new Array(n).fill(0);
    ext[i] = 1;
    return [...row, ...ext];
  });

  for (let col = 0; col < n; col++) {
    let pivotRow = col;
    let pivotVal = Math.abs(aug[col][col]);
    for (let row = col + 1; row < n; row++) {
      if (Math.abs(aug[row][col]) > pivotVal) {
        pivotRow = row;
        pivotVal = Math.abs(aug[row][col]);
      }
    }
    if (pivotVal < 1e-12) return null;

    if (pivotRow !== col) {
      [aug[col], aug[pivotRow]] = [aug[pivotRow], aug[col]];
    }

    const pivot = aug[col][col];
    for (let j = 0; j < 2 * n; j++) aug[col][j] /= pivot;

    for (let row = 0; row < n; row++) {
      if (row === col) continue;
      const factor = aug[row][col];
      for (let j = 0; j < 2 * n; j++) {
        aug[row][j] -= factor * aug[col][j];
      }
    }
  }

  return aug.map((row) => row.slice(n));
}
